go/vt/vtgate/evalengine/string.go

```diff
@@ -0,0 +1,125 @@
+/*
+Copyright 2022 The Vitess Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package evalengine
+
+import (
+	"strconv"
+	"unicode"
+
+	"vitess.io/vitess/go/mysql/collations"
+	"vitess.io/vitess/go/sqltypes"
+	vtrpcpb "vitess.io/vitess/go/vt/proto/vtrpc"
+	"vitess.io/vitess/go/vt/vterrors"
+)
+
+type (
+	builtinAscii     struct{}
+	builtinBin       struct{}
+	builtinBitLength struct{}
+)
+
+func (builtinAscii) call(_ *ExpressionEnv, args []EvalResult, result *EvalResult) {
+	toascii := &args[0]
+	if toascii.isNull() {
+		result.setNull()
+		return
+	}
+
+	toascii.makeBinary()
+	if len(toascii.bytes()) > 0 {
+		result.setInt64(int64(toascii.bytes()[0]))
+	} else {
+		result.setInt64(0)
+	}
+}
+
+func (builtinAscii) typeof(env *ExpressionEnv, args []Expr) (sqltypes.Type, flag) {
+	if len(args) != 1 {
+		throwArgError("ASCII")
+	}
+	_, f := args[0].typeof(env)
+	return sqltypes.Int64, f
+}
+
+func (builtinBin) call(env *ExpressionEnv, args []EvalResult, result *EvalResult) {
+	tobin := &args[0]
+	if tobin.isNull() {
+		result.setNull()
+		return
+	}
+
+	var raw []byte
+	switch t := tobin.typeof(); {
+	case sqltypes.IsNumber(t):
+		tobin.makeUnsignedIntegral()
+		raw = strconv.AppendUint(raw, tobin.uint64(), 2)
+	case sqltypes.IsText(t):
+		input := tobin.string()
+		index := len(input)
+		if index == 0 {
+			break
+		}
+		for i, ch := range input {
+			if !unicode.IsDigit(ch) {
+				index = i
+				break
+			}
+		}
+		intstr := input[:index]
+		toint64, _ := strconv.ParseInt(intstr, 10, 64)
+		raw = strconv.AppendUint(raw, uint64(toint64), 2)
```

Comment:
This doesn't look right. We already have a function that tries to parse numbers into `float` and handles this logic properly. I would try using that one, and then converting the `float64` to `uint64`. I have a hunch that's what MySQL is doing behind the scenes.

Reply:
Thanks for your review!

In my consideration, the above code is mainly to deal with strings containing both letters and numbers, not only for float. Such as "123abc".

I tried to write a piece of code, firstly convert to float and then convert to uint64. Like this:

```go
func (builtinBin) call(env *ExpressionEnv, args []EvalResult, result *EvalResult) {
	tobin := &args[0]
	if tobin.isNull() {
		result.setNull()
		return
	}

	var raw []byte
	switch t := tobin.typeof(); {
	case sqltypes.IsNumber(t):
		tobin.makeUnsignedIntegral()
		raw = strconv.AppendUint(raw, tobin.uint64(), 2)
	case sqltypes.IsText(t):
		if len(tobin.bytes()) == 0 {
			return
		}
		tobin.makeFloat()
		float, _ := tobin.coerceToFloat()
		raw = strconv.AppendUint(raw, uint64(float), 2)
	default:
		throwEvalError(vterrors.Errorf(vtrpcpb.Code_UNIMPLEMENTED, "Unsupported BIN argument: %s", t.String()))
	}

	result.setRaw(sqltypes.VarChar, raw, collations.TypedCollation{
		Collation:    env.DefaultCollation,
		Coercibility: collations.CoerceCoercible,
		Repertoire:   collations.RepertoireASCII,
	})
}
```

It was well behaved when inputted ”123abc“ or "123.456".
But for strings like "2e2", it has a difference with MySQL's output.
```shell
mysql> select bin("2e2");
+------------+
| bin(:vtg1) |
+------------+
| 11001000   |
+------------+
1 row in set (1.42 sec)
```

While the MySQL5.7 returned `10`.
```shell
mysql> select bin("2e2");
+------------+
| bin("2e2") |
+------------+
| 10         |
+------------+
1 row in set (0.00 sec)
```

Is it possible to bypass string processing?
I'm not very familiar with go code and would like some advice from you, thanks!